ds3-sdk/src/main/java/com/spectralogic/ds3client/metadata/MetadataReceivedListenerImpl.java

```diff
@@ -0,0 +1,254 @@
+
+        if (storedOS != null && storedOS.equals(os) && os.contains("Windows")) {
+            setPermissionsForWindows(metadata, objectName);
+        } else {
+            if (metadata.get("ds3-permisions").size() > 0) {
+                String permissions = metadata.get("ds3-permisions").get(0);
+                if (permissions != null && !permissions.equals("")) {
+                    final Matcher m = Pattern.compile("\\(([^)]+)\\)").matcher(permissions);
+                    while (m.find()) {
+                        permissions = (m.group(1));
+                    }
+                    setPermissions(localFilePath + "/" + objectName, permissions);
+                }
+            }
+        }
+
+
+        for (final String metadataKey : metadata.keys()) {
+            final List<String> values = metadata.get(metadataKey);
+            builder.append("<Key: ")
+                    .append(metadataKey)
+                    .append(" Values: ")
+                    .append(joiner.join(values))
+                    .append("> ");
+        }
+    }
+
+    private void setPermissionsForWindows(final Metadata metadata, final String objectName) {
+        final Path path = Paths.get(localFilePath + "/" + objectName);
+        final AclFileAttributeView view1 = Files.getFileAttributeView(path, AclFileAttributeView.class);
+        final List<AclEntry> aclEntries = new ArrayList<AclEntry>();
+
+        if (metadata.get("ds3-ownerPerm").size() > 0) {
+            String ownerPermission = metadata.get("ds3-ownerPerm").get(0);
+            restorePermissionByUser(ownerPermission, "Administrators", path, aclEntries);
+        }
+        if (metadata.get("ds3-authenticatedUser").size() > 0) {
+            String authenticatedUser = metadata.get("ds3-authenticatedUser").get(0);
+            restorePermissionByUser(authenticatedUser, "Authenticated Users", path, aclEntries);
+        }
+        if (metadata.get("ds3-systemUser").size() > 0) {
+            String systemUserPermission = metadata.get("ds3-systemUser").get(0);
+            restorePermissionByUser(systemUserPermission, "SYSTEM", path, aclEntries);
+        }
+        if (metadata.get("ds3-allUser").size() > 0) {
+            String allUserPermission = metadata.get("ds3-allUser").get(0);
+            restorePermissionByUser(allUserPermission, "Users", path, aclEntries);
+        }
+        try {
+            view1.setAcl(aclEntries);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void restorePermissionByUser(final String permission, final String user, final Path path, final List<AclEntry> aclEntries) {
+        try {
+            final AclEntry.Builder builderWindow = AclEntry.newBuilder();
+            final UserPrincipal bRiceUser = FileSystems.getDefault()
+                    .getUserPrincipalLookupService().lookupPrincipalByName(user);
+            final Set<AclEntryPermission> permissions = new HashSet<AclEntryPermission>();
```

Comment:
Update this to use a Guava ImmutableSet.Builder<AclEntryPermission> and then on line 187 build it and pass that into the builderWindow